### PR TITLE
test: Use a more robust way to break Insights uploads 

### PR DESCRIPTION
**Pull request by mvollmer**
Insights-client seems to have gotten slower recently, so let's wait longer for it to fail the upload.

And also, provoke the upload failure by returning a proper error code from mock-insights.

### CODE REVIEW

**Comment by martinpitt** [comment]
@mvollmer Note the "black" failure

**Comment by mvollmer** [comment]
> @mvollmer Note the "black" failure

Fixed.

**Comment by martinpitt** [comment]
@ptoscano can you please have a look?

File: test/check-subscriptions [hunk 1]
```diff
@@ -380,7 +380,8 @@ class TestSubscriptions(SubscriptionsCase):
 
         dialog_register_button_sel = "footer .pf-m-primary"
         b.click(dialog_register_button_sel)
-        b.wait_not_present(dialog_register_button_sel)
+        with b.wait_timeout(360):
+            b.wait_not_present(dialog_register_button_sel)
```
Comment (ptoscano) [inline test/check-subscriptions:384]: hmm the code here is the same as used in other tests, and in this case the connection to insights is disabled when registering; how come this is needed?

File: test/check-subscriptions [hunk 2]
```diff
@@ -440,8 +441,17 @@ class TestSubscriptions(SubscriptionsCase):
...
+        m.execute(
+            ["sed", "-i", "-e", "s/--retry 3/--retry 1/", "/usr/lib/systemd/system/insights-client.service"]
+        )
+        m.execute(["systemctl", "daemon-reload"])
+        m.execute(["touch", "/etc/insights-client/mock-upload-fail"])
+        m.execute(
+            "systemctl start insights-client; "
+            "while systemctl --quiet is-active insights-client; do sleep 1; done",
+            timeout=360,
+        )
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply test/check-subscriptions:454]: Oh, I would say this is actually less playing with insights-client.service. The `mock-upload-fail` file is read by our own mock-insights server and causes it to return a error code during uploading.

I think the machin-id.lost method actually didn't cause a failure anymore, insights-client would just recreate that file, but I am not sure.

File: test/check-subscriptions (hunk 1 above)
Comment (mvollmer) [reply test/check-subscriptions:384]: Hmm, maybe registering was actually the problem, not connecting to insights. I'll check this more carefully by comparing old and new images.

Comment (martinpitt) [reply test/check-subscriptions:384]: Note that this didn't "binary" break on an image refresh. It has been flaky for months, and just recently got bad enough that it pretty much always fails. But there's no "known good state" to compare it to.

File: test/check-subscriptions (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply test/check-subscriptions:454]: Ah, or do you mean the sedery re `--retry 3`?

File: test/check-subscriptions (hunk 1 above)
Comment (mvollmer) [reply test/check-subscriptions:384]: Alright, without the increased timeout, we get this warning:

```
WARNING: Waiting for !ph_is_present("footer .pf-m-primary") took 10.8 seconds, which is 72% of the timeout.
```

That's probably why I made it longer.

There is a similar warning for `testRegister`:

```
WARNING: Waiting for !ph_in_text("body","User doc is member of more organizations, but no organization was selected") took 7.7 seconds, which is 51% of the timeout.
```

So registration seems to brush up against the timeout in general.  I'll make these two places wait longer, in a separate commit.

Comment (mvollmer) [reply test/check-subscriptions:388]: Done.

File: test/check-subscriptions (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply test/check-subscriptions]: Ok, removing machine-id no longer breaks insights-client. I agree that modifying the unit file of insights-client.service is questionable, so let's not do it. The price is that we need to wait 6 minutes longer for the failure to appear... :-/

**Comment by mvollmer** [comment]
@ptoscano, thanks for not letting me get away with such a sloppy explanation! :-)

PTAL

**Comment by ptoscano** [comment]
Ah, and I think I know why the "break insights-client" way does not work anymore: because of other changes (see CCT-161 in jira), the generation of the `machine-id` file of `insights-client` now uses the UUID of `subscription-manager` in case the system is registered with it (see https://github.com/RedHatInsights/insights-core/pull/4057); this new behaviour is in insights-core 3.3.16, available for any system since March 26th/27th. The sequence of the events is thus the following:
- the system is registered with `subscription-manager`
- when registering with `insights-client`, `insights-client` generates a new `machine-id` file based on the `subscription-manager` UUID
- the "mock insights" registers the system using that UUID
- the `machine-id` file is moved away to `machine-id.lost`
- `insights-client` is run to get the status: `insights-client` generates a new one as it thinks it is registered (see the `/etc/insights-client/.registered` file), and since the UUID is the same as `subscription-manager`, then it will find itself in the "mock insights", and consider the situation as correct, keeping the newly "regenerated" `machine-id` file

Have you tried to move away both the `machine-id` and `.registered` files?

**Comment by ptoscano** [comment]
> Have you tried to move away both the `machine-id` and `.registered` files?

Or maybe putting in `machine-id` a newly generated UUID, so that the "mock insights" returns a failure... which it does not ATM :-(

**Comment by ptoscano** [comment]
Regarding the first commit, i.e. `test: Wait longer for registering`: from what I understand, the default timeout is 15 seconds, and this commit is bumping it for some operations to 360 seconds (i.e. 6 minutes); isn't the bump a bit too much, and maybe use something less loose such as 60 seconds?

Otherwise, the commit seems fine for me.

**Comment by mvollmer** [comment]
> Regarding the first commit, i.e. `test: Wait longer for registering`: from what I understand, the default timeout is 15 seconds, and this commit is bumping it for some operations to 360 seconds (i.e. 6 minutes); isn't the bump a bit too much, and maybe use something less loose such as 60 seconds?

Sure.

**Comment by mvollmer** [comment]
> > Have you tried to move away both the `machine-id` and `.registered` files?
> 
> Or maybe putting in `machine-id` a newly generated UUID, so that the "mock insights" returns a failure... which it does not do ATM :-(

No, I have not tried that. Isn't it best to return a error code from mock-insights? That is the most realistic thing to do, no?

The downside is that the test is now really slow. Does that matter much?

**Comment by ptoscano** [comment]
> > > Have you tried to move away both the `machine-id` and `.registered` files?
> > 
> > 
> > Or maybe putting in `machine-id` a newly generated UUID, so that the "mock insights" returns a failure... which it does not do ATM :-(
> 
> No, I have not tried that. Isn't it best to return a error code from mock-insights? That is the most realistic thing to do, no?

Yeah, that's what I think it ought to be done, and thus `mock-insights` needs to be fixed.

> The downside is that the test is now really slow. Does that matter much?

I think that doing the following steps:
- fixing `mock-insights` to actually error out for a non-registered system
- "breaking" `insights-client` by putting a wrong UUID (and restoring it afterwards, of course)

should hopefully "fix" the test as it was supposed to do (i.e. fail as expected).

File: test/check-subscriptions [hunk 3]
```diff
@@ -380,7 +383,8 @@ class TestSubscriptions(SubscriptionsCase):
 
         dialog_register_button_sel = "footer .pf-m-primary"
         b.click(dialog_register_button_sel)
-        b.wait_not_present(dialog_register_button_sel)
+        with b.wait_timeout(360):
```
Comment (ptoscano) [inline test/check-subscriptions:386]: `60` here too?

**Comment by ptoscano** [comment]
@mvollmer I took the liberty to take some of your changes (the timeout increases), and create a new PR with what I think should be a better approach; please check #66 when you have the possibility :)

**Comment by ptoscano** [comment]
Obsoleted by #66 (which contains parts of your work here, @mvollmer ) -- thanks!

**Pull request closed without merging**